File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void backtrack_comb(int start, int target, vector<int>& candidates, vector<int>& arr, vector<vector<int>>& result){
        if (target == 0) {
            result.push_back(arr);
            return;
        }
        for (int idx=start; idx < candidates.size(); idx++) {
            if (candidates[idx] <= target) {
                arr.push_back(candidates[idx]);
                backtrack_comb(idx, target-candidates[idx], candidates, arr, result);
                arr.pop_back();
            }
        }

    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> arr;
        backtrack_comb(0, target, candidates, arr, result);
        return result;
        
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
  
        def backtrack_comb(arr, result, target) :
            if target == 0 and sorted(arr[:]) not in result :
                result.append(sorted(arr[:]))
                return
            for i in range(len(candidates)) :
                if candidates[i] <= target :
                    arr.append(candidates[i])
                    backtrack_comb(arr, result, target-candidates[i])
                    arr.pop()

        result = []
        backtrack_comb([], result, target)
        return list(result)